numpy/lib/function_base.py

```diff
@@ -3929,21 +3929,32 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        diff_x2_x1 = x2 - x1
+        r_above = x1 + diff_x2_x1 * weights_above
+        r_below = x2 - diff_x2_x1 * weights_below
```

Comment:
You should probably add an inline comment or a docstring comment that this implementation was suggested by Pedro Gimeno at https://math.stackexchange.com/a/1798323/383305 to ensure monotonicity with floating point operations.